### DISCUSSION TITLE
How to install Rosetta homebrew to /usr/local on Apple Silicon

### DISCUSSION THREAD

**Discussion by mikepqr** [Homebrew/discussions#885] [Getting started] [answered]
Homebrew installs in to /opt/homebrew on Apple Silicon. This worked for me and all is fine.

I would _also_ like to install a separate Rosetta installation into /usr/local. (In my case, this is so I can install shellcheck from a bottle.)

How do I actually do this? The release notes say "Rosetta 2 on Apple Silicon still provides support for Intel x86_64 in /usr/local". Sounds great! And the [installation instructions](https://docs.brew.sh/Installation) emphasize all the places you should _not_ install homebrew. Also sounds great! But they leave out the crucial piece of information: how to install it in the _right_ place.

Do I do `mkdir -p /usr/local && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C /usr/local`?

I'm reluctant to just give this a try because I have other stuff in /usr/local.

**Comment by Bo98** [comment, marked as answer]
If you launch Terminal in Rosetta mode then the installer should work automatically.

Alternatively, you can prefix commands with `arch -x86_64`.

**Reply by mikepqr** [reply]
By "the installer should work automatically" do you mean it should install homebrew in /usr/local as it does on an Intel Mac? That's what I'm trying to do, but that's not the behavior I see:

```
$ arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
==> Checking for `sudo` access (which may request your password).
Password:
==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew
stat: cannot read file system information for '%A': No such file or directory
stat: cannot read file system information for '%A': No such file or directory
stat: cannot read file system information for '%u': No such file or directory
stat: cannot read file system information for '%g': No such file or directory
stat: cannot read file system information for '%u': No such file or directory
stat: cannot read file system information for '%g': No such file or directory
==> The following existing directories will be made writable by user only:
/opt/homebrew/share/zsh
/opt/homebrew/share/zsh/site-functions
==> The following existing directories will have their owner set to mike:
/opt/homebrew/share/zsh
/opt/homebrew/share/zsh/site-functions
==> The following existing directories will have their group set to admin:
/opt/homebrew/share/zsh
/opt/homebrew/share/zsh/site-functions

Press RETURN to continue or any other key to abort
```

**Reply by Bo98** [reply]
The `arch` command sometimes does not do a good job at telling you if something's wrong. 

Do you have Rosetta 2 installed? It isn't installed by default.

**Reply by mikepqr** [reply]
Thank you for the quick replies! I am 99% certain I accepted a dialog to install it, and there is an "Open In Rosetta" checkbox option on Terminal.app that I can optionally select, so yes I think I have it installed.

This is what I get if I launch Terminal.app in Rosetta mode (note `arch` on the first line):

```
$ arch
i386
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
==> Checking for `sudo` access (which may request your password).
Password:
==> This script will install:
/opt/homebrew/bin/brew
/opt/homebrew/share/doc/homebrew
/opt/homebrew/share/man/man1/brew.1
/opt/homebrew/share/zsh/site-functions/_brew
/opt/homebrew/etc/bash_completion.d/brew
/opt/homebrew
...
```

What is the _expected_ behavior here? Is the install script supposed to install in /usr/local if the output of `arch` is i386?

And will the command I suggested in my original post help (`mkdir -p /usr/local && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C /usr/local`)?

**Reply by Bo98** [reply]
> Is the install script supposed to install in /usr/local if the output of `arch` is i386?

Sort of, yes. The actual value checked is `uname -m`.

**Reply by mikepqr** [reply]
Well that's the problem then!
```
$ uname -m
arm64
$ arch
i386
```
Any ideas?

**Reply by mikepqr** [reply]
Ah. Got it. I put arm64 coreutils in my path:
```
$ file $(which uname)
/opt/homebrew/opt/coreutils/libexec/gnubin/uname: Mach-O 64-bit executable arm64
```

**Reply by Bo98** [reply]
Good spot! We should probably fix the install script to use /usr/bin/uname directly.

**Reply by mikepqr** [reply]
That sounds like a good idea. 

I kind of wonder whether `uname` should be treated as a special case in the `coreutils` package too? If it can't be shipped as a universal binary, perhaps it should be removed from the package, given that it's guaranteed to give incorrect output in some circumstances. Is this worth filing a ticket on the main homebrew repo?

**Reply by Bo98** [reply]
I think the issue comes from system `uname` being a universal binary but `coreutils` being only arm64. When there's multiple slices to choose from, the system will pick the one matching the arch mode you are in. But there isn't an x86_64 slice in the `coreutils` binary so it'll just always print `arm64` (`uname -m` does nothing magical or dynamic - it's a baked-in constant at compile time).

Basically a collection of "intentional behaviour" scenarios clashing with each other, so there's no real "bug" anywhere.

**Reply by mikepqr** [reply]
Yup. Which means brew ships a `uname` which is essentially a hard-coded print statement, i.e. fundamentally broken on a computer that is de facto dual architecture. Which is why I wonder if it should be removed if there's no plans for universal binaries.

Anyway, do you want me to open a ticket to hard-code /usr/bin/uname? The PR is beyond me, but the feature request seems uncontroversial.

**Reply by Bo98** [reply]
Sure. https://github.com/Homebrew/install will be the place for that.

**Reply by mikepqr** [reply]
Thanks for your help! I opened a couple of tickets.

https://github.com/Homebrew/brew/issues/10676
https://github.com/Homebrew/install/issues/531

**Reply by mikepqr** [reply]
And thanks for moving the 10676 to the right repository!

**Reply by owolabiezekiel** [reply]
> Ah. Got it. I put arm64 coreutils in my path:
> 
> ```
> $ file $(which uname)
> /opt/homebrew/opt/coreutils/libexec/gnubin/uname: Mach-O 64-bit executable arm64
> ```

How did you do this. Can go give a detailed step by step guide on how to achieve this? I haven't been able to set it up successfully
